src/components/Session/withClowdrAppState.tsx

```diff
@@ -1,22 +1,121 @@
-    // (and maybe ClowdrAppState can be globally renamed just ClowdrState...?)
-    class WithAuthentication extends React.Component {
-
-        constructor(props) {
+import { RoomInvalidParametersError } from 'twilio-video';
+import BreakoutRoom from '../../classes/BreakoutRoom';
+import UserProfile from '../../classes/UserProfile';
+import UserPresence from '../../classes/UserPresence';
+import SocialSpace from '../../classes/SocialSpace';
+import ClowdrInstance from '../../classes/ClowdrInstance';
+import LiveActivity from '../../classes/LiveActivity';
+
+interface WithAuthenticationProps {
+    history: string[];   
+} 
+
+interface FlairUIData {color: string, tooltip: string}
+interface AllFlairData {value: string, color: string, tooltip: string, id: string, priority: number}
+
+interface WithAuthenticationState {
+    // TS: Fill in the types!
+    user: any,
+    users: any,
+    loading: boolean,
+    roles: any,
+    currentRoom: any,
+    history: string[] | undefined,
+    refreshUser: any,
+    getChatClient: any,
+    setSocialSpace: any,
+    getConferenceBySlackName: any,
+    setActiveRoom: any,
+    currentConference: ClowdrInstance | undefined | null,  // TS: Really??
+    activeRoom: any,
+    helpers: any,
+    chatClient: any,
+    parseLive: any,
+    presences: any,
+    videoRoomsLoaded: boolean,
+    liveVideoRoomMembers: any,
+    activePublicVideoRooms: BreakoutRoom[],
+    activePrivateVideoRooms: BreakoutRoom[],
+    userProfile: UserProfile | null,  // TS: Or should it be undefined??
+    permissions: any,
+    leftSidebar: any,
+    activeSpace: any,
+    spaces: any,
+    chatChannel: any,
+    programCache: any,
+    isAdmin: boolean;
+    isModerator: boolean;
+    isManager: boolean;
+    isClowdrAdmin: boolean;
+    flairColors: Record<string,FlairUIData>;
+    allFlair: AllFlairData[];
+}
+
+type RoomID = string    // TS: Doesn't belong here?
+type UserID = string    // TS: Doesn't belong here?  And should this be the same as the next??
+type UserProfileID = string    // TS: Doesn't belong here?
+type Subscriber = React.Component;   // TS: What arguments?
+
+// TS: This should be an enumeration of strings -- get all the possibilities from the PrivilegedAction table in the DB
+type Permission = string
+
+// TS: This belongs in some top-level utility module
+function assert(input: any): asserts input { // <-- not a typo
+    if (!input) throw new Error('Assertion failed!');
+}
+
+const withAuthentication = (Component: React.Component<WithAuthenticationProps, WithAuthenticationState>) => {
+    // @Jon/@Crista/@Benjamin    (maybe ClowdrAppState can be globally renamed just ClowdrState...?)
+    class WithAuthentication extends React.Component<WithAuthenticationProps, WithAuthenticationState> {
+        // TS: @Jon - I (BCP) don't understand the logic here very well (Crista doesn't either).  Why do we initialize all these fields of "this" and then copy most of them to this.state??
```

Comment:
When we first implemented this, we pushed updates from live queries into sub-components by modifying the `state` of this context object. Calling `setState` on the `withAuthentication` will result in every component on the page being re-rendered. This is not a good patten. I have been trying to refactor things away so that mostly what is in `state` are helper methods to expose to components, and any of the actual data is stored as fields of `this`.